File: src/Client/TerrainAtlas.c
#include "TerrainAtlas.h"
#include "Bitmap.h"
#include "Block.h"
#include "ExtMath.h"
#include "Funcs.h"
#include "GraphicsAPI.h"
#include "Platform.h"

void Atlas2D_UpdateState(Bitmap* bmp) {
	Atlas2D_Bitmap = *bmp;
	Atlas2D_TileSize = bmp->Width / ATLAS2D_TILES_PER_ROW;
	Block_RecalculateSpriteBB();
}

static GfxResourceID Atlas2D_LoadTextureElement_Raw(TextureLoc texLoc, Bitmap* element) {
	Int32 size = Atlas2D_TileSize;
	Int32 x = Atlas2D_TileX(texLoc), y = Atlas2D_TileY(texLoc);
	Bitmap_CopyBlock(x * size, y * size, 0, 0, &Atlas2D_Bitmap, element, size);

	return Gfx_CreateTexture(element, false, Gfx_Mipmaps);
}

GfxResourceID Atlas2D_LoadTile(TextureLoc texLoc) {
	Int32 size = Atlas2D_TileSize;
	Bitmap tile;

	/* Try to allocate bitmap on stack if possible */
	if (size > 64) {
		Bitmap_Allocate(&tile, size, size);
		GfxResourceID texId = Atlas2D_LoadTextureElement_Raw(texLoc, &tile);
		Platform_MemFree(&tile.Scan0);
		return texId;
	} else {
		UInt8 scan0[Bitmap_DataSize(64, 64)];
		Bitmap_Create(&tile, size, size, scan0);
		return Atlas2D_LoadTextureElement_Raw(texLoc, &tile);
	}
}

void Atlas2D_Free(void) {
	Platform_MemFree(&Atlas2D_Bitmap.Scan0);
}


TextureRec Atlas1D_TexRec(TextureLoc texLoc, Int32 uCount, Int32* index) {
	*index  = Atlas1D_Index(texLoc);
	Int32 y = Atlas1D_RowId(texLoc);

	/* Adjust coords to be slightly inside - fixes issues with AMD/ATI cards. */
	TextureRec rec;
	rec.U1 = 0.0f; 
	rec.V1 = y * Atlas1D_InvTileSize;
	rec.U2 = (uCount - 1) + UV2_Scale;
	rec.V2 = rec.V1 + UV2_Scale * Atlas1D_InvTileSize;
	return rec;
}

static void Atlas1D_Make1DTexture(Int32 i, Int32 atlas1DHeight, Int32* index) {
	Int32 tileSize = Atlas2D_TileSize;
	Bitmap atlas1D;
	Bitmap_Allocate(&atlas1D, tileSize, atlas1DHeight);

	Int32 index1D;
	for (index1D = 0; index1D < Atlas1D_TilesPerAtlas; index1D++) {
		Int32 atlasX = (*index % ATLAS2D_TILES_PER_ROW) * tileSize;
		Int32 atlasY = (*index / ATLAS2D_TILES_PER_ROW) * tileSize;

		Bitmap_CopyBlock(atlasX, atlasY, 0, index1D * tileSize,
			&Atlas2D_Bitmap, &atlas1D, tileSize);
		(*index)++;
	}

	Atlas1D_TexIds[i] = Gfx_CreateTexture(&atlas1D, true, Gfx_Mipmaps);
	Platform_MemFree(&atlas1D.Scan0);
}

static void Atlas1D_Convert2DTo1D(Int32 atlasesCount, Int32 atlas1DHeight) {
	Atlas1D_Count = atlasesCount;
	Platform_Log2("Loaded new atlas: %i bmps, %i per bmp", &atlasesCount, &Atlas1D_TilesPerAtlas);

	Int32 index = 0, i;
	for (i = 0; i < atlasesCount; i++) {
		Atlas1D_Make1DTexture(i, atlas1DHeight, &index);
	}
}

void Atlas1D_UpdateState(void) {
	Int32 maxAtlasHeight = min(4096, Gfx_MaxTextureDimensions);
	Int32 maxTilesPerAtlas = maxAtlasHeight / Atlas2D_TileSize;
	Int32 maxTiles = ATLAS2D_ROWS_COUNT * ATLAS2D_TILES_PER_ROW;

	Atlas1D_TilesPerAtlas = min(maxTilesPerAtlas, maxTiles);
	Int32 atlasesCount = Math_CeilDiv(maxTiles, Atlas1D_TilesPerAtlas);
	Int32 atlasHeight = Atlas1D_TilesPerAtlas * Atlas2D_TileSize;

	Atlas1D_InvTileSize = 1.0f / Atlas1D_TilesPerAtlas;
	Atlas1D_Convert2DTo1D(atlasesCount, atlasHeight);
}

Int32 Atlas1D_UsedAtlasesCount(void) {
	TextureLoc maxTexLoc = 0;
	Int32 i;

	for (i = 0; i < Array_Elems(Block_Textures); i++) {
		maxTexLoc = max(maxTexLoc, Block_Textures[i]);
	}
	return Atlas1D_Index(maxTexLoc) + 1;
}

void Atlas1D_Free(void) {
	Int32 i;
	for (i = 0; i < Atlas1D_Count; i++) {
		Gfx_DeleteTexture(&Atlas1D_TexIds[i]);
	}
}

File: ClassicalSharp/ClassicalSharp/TexturePack/TerrainAtlas.cs
// Copyright 2014-2017 ClassicalSharp | Licensed under BSD-3
using System;
using System.Drawing;
using ClassicalSharp.GraphicsAPI;
#if ANDROID
using Android.Graphics;
#endif

namespace ClassicalSharp.Textures {
	
	/// <summary> Represents a 2D packed texture atlas, specifically for terrain.png. </summary>
	public static class Atlas2D {		
		public const int TilesPerRow = 16, MaxRowsCount = 32;
		public static Bitmap Atlas;
		public static int TileSize, RowsCount;
		internal static Game game;

		public static void UpdateState(Bitmap bmp) {
			Atlas = bmp;
			TileSize = bmp.Width / TilesPerRow;
			RowsCount = bmp.Height / TileSize;
			RowsCount = Math.Min(RowsCount, MaxRowsCount);
			BlockInfo.RecalculateSpriteBB();
		}

		public static int LoadTile(int texLoc) {
			int size = TileSize, x = texLoc % TilesPerRow, y = texLoc / TilesPerRow;
			if (y >= RowsCount) return 0;
			
			using (FastBitmap atlas = new FastBitmap(Atlas, true, true))
				using (Bitmap bmp = Platform.CreateBmp(size, size))
					using (FastBitmap dst = new FastBitmap(bmp, true, false))
			{
				FastBitmap.MovePortion(x * size, y * size, 0, 0, atlas, dst, size);
				return game.Graphics.CreateTexture(dst, false, game.Graphics.Mipmaps);
			}
		}

		public static void Dispose() {
			if (Atlas != null) Atlas.Dispose();
		}
	}
	
	/// <summary> Represents a 2D packed texture atlas that has been converted into an array of 1D atlases. </summary>
	public static class Atlas1D {
		public static int TilesPerAtlas, AtlasesCount, Mask, Shift;
		public static float invTileSize;		
		internal static Game game;
		public const int MaxAtlases = Atlas2D.MaxRowsCount * Atlas2D.TilesPerRow;
		public static int[] TexIds = new int[MaxAtlases];
		
		public static TextureRec GetTexRec(int texLoc, int uCount, out int index) {
			index = texLoc >> Shift;
			int y = texLoc &  Mask;
			// Adjust coords to be slightly inside - fixes issues with AMD/ATI cards.
			return new TextureRec(0, y * invTileSize, (uCount - 1) + 15.99f/16f, (15.99f/16f) * invTileSize);
		}
		
		/// <summary> Returns the index of the 1D texture within the array of 1D textures
		/// containing the given texture id. </summary>
		public static int Get1DIndex(int texLoc) { return texLoc >> Shift; } // texLoc / TilesPerAtlas;
		
		/// <summary> Returns the index of the given texture id within a 1D texture. </summary>
		public static int Get1DRowId(int texLoc) { return texLoc &  Mask;  } // texLoc % TilesPerAtlas
		
		public static void UpdateState() {
			int maxAtlasHeight   = Math.Min(4096, game.Graphics.MaxTexHeight);
			int maxTilesPerAtlas = maxAtlasHeight / Atlas2D.TileSize;
			int maxTiles         = Atlas2D.RowsCount * Atlas2D.TilesPerRow;

			TilesPerAtlas   = Math.Min(maxTilesPerAtlas, maxTiles);
			AtlasesCount    = Utils.CeilDiv(maxTiles, TilesPerAtlas);
			int atlasHeight = TilesPerAtlas * Atlas2D.TileSize;

			invTileSize = 1.0f / TilesPerAtlas;
			Mask        = TilesPerAtlas - 1;
			Shift       = Utils.Log2(TilesPerAtlas);

			Utils.LogDebug("Loaded new atlas: {0} bmps, {1} per bmp", AtlasesCount, TilesPerAtlas);
			int index = 0;			
			using (FastBitmap atlas = new FastBitmap(Atlas2D.Atlas, true, true)) {
				for (int i = 0; i < AtlasesCount; i++) {
					Make1DTexture(i, atlas, atlasHeight, ref index);
				}
			}
		}
		
		static void Make1DTexture(int i, FastBitmap atlas, int atlas1DHeight, ref int index) {
			int tileSize = Atlas2D.TileSize;
			using (Bitmap atlas1d = Platform.CreateBmp(tileSize, atlas1DHeight))
				using (FastBitmap dst = new FastBitmap(atlas1d, true, false))
			{
				for (int index1D = 0; index1D < TilesPerAtlas; index1D++) {
					int atlasX = (index % Atlas2D.TilesPerRow) * tileSize;
					int atlasY = (index / Atlas2D.TilesPerRow) * tileSize;
					
					FastBitmap.MovePortion(atlasX, atlasY,
					                       0, index1D * tileSize, atlas, dst, tileSize);
					index++;
				}
				TexIds[i] = game.Graphics.CreateTexture(dst, true, game.Graphics.Mipmaps);
			}
		}
		
		public static int UsedAtlasesCount() {
			int maxTexLoc = 0;
			for (int i = 0; i < BlockInfo.textures.Length; i++) {
				maxTexLoc = Math.Max(maxTexLoc, BlockInfo.textures[i]);
			}			
			return Get1DIndex(maxTexLoc) + 1;
		}
		
		public static void Dispose() {
			if (TexIds == null) return;
			
			for (int i = 0; i < AtlasesCount; i++) {
				game.Graphics.DeleteTexture(ref TexIds[i]);
			}
		}
	}
}